src/unauthenticated_stream/mod.rs

```diff
@@ -0,0 +1,182 @@
+// Copyright 2018 Brian Smith.
+//
+// Permission to use, copy, modify, and/or distribute this software for any
+// purpose with or without fee is hereby granted, provided that the above
+// copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHORS DISCLAIM ALL WARRANTIES
+// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY
+// SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
+// OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+// CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+//! Unauthenticated stream cipher.
+//!
+//! # Warning
+//! This API does not provide any authentification! You should use the
+//! [aead](aead/index.html) module instead.
+//!
+//! C analog: `EVP_CIPHER`
+//!
+//! Go analog: [`crypto.cipher.Stream`]
+//!
+//! [`crypto.cipher.Stream`]: https://golang.org/pkg/crypto/cipher/#Stream
+
+use {chacha, error, init, polyfill};
+
+pub use self::chacha20::CHACHA20;
+
+/// A key for an unauthenticated stream cipher.
+///
+/// C analog: `EVP_CIPHER_CTX`
+///
+/// Go analog: [`crypto.cipher.Stream`]
+pub struct StreamingKey {
+    key: Key,
+}
+
+impl StreamingKey {
+    /// Create a new streaming key.
+    ///
+    /// `key_bytes` must be exactly `algorithm.key_len` bytes long.
+    ///
+    /// C analogs: `EVP_CIPHER_CTX_init`.
+    ///
+    /// Go analog:
+    ///   [`crypto.aes.NewCipher`](https://golang.org/pkg/crypto/aes/#NewCipher)
+    #[inline]
+    pub fn new(algorithm: &'static Algorithm, key_bytes: &[u8])
+               -> Result<StreamingKey, error::Unspecified> {
+        Ok(StreamingKey {
+            key: Key::new(algorithm, key_bytes)?,
+        })
+    }
+
+    /// The key's unauthenticated stream cipher algorithm.
+    ///
+    /// C analog: `EVP_CIPHER_CTX.cipher`
+    #[inline(always)]
+    pub fn algorithm(&self) -> &'static Algorithm { self.key.algorithm() }
+}
+
+/// Applies the keystream to the input data in place.
+pub fn xor_keystream_in_place<'a>(key: &StreamingKey, nonce: &[u8],
```

Comment:
I thought in the discussion we agreed that we didn't need an in-place API. I'd rather skip the in-place API and provide an out-of-place API with separate in and out parameters unless the people implementing QUIC think in-place is better. (If in-place is better, why?)

Assuming we have separate encryption key and decryption key types, then we'd also have separate encrypt and decrypt functions.

I don't think we need `keystream` in the name. I find that to be confusing.

Reply:
> I thought in the discussion we agreed that we didn't need an in-place API. I'd rather skip the in-place API and provide an out-of-place API with separate in and out parameters unless the people implementing QUIC think in-place is better. (If in-place is better, why?)

Ah, sorry, I forgot to mention this in the description. I went back to the in-place implementation because it makes things a little bit better for QUIC implementations. The problem is that the QUIC packet header, including the _unencrypted_ packet number, is used as AD for the decryption of the main packet payload. So an implementation needs to first decrypt the packet number, write the packet number plaintext back into the packet header buffer, and then decrypt the payload by splitting the buffer between header (used as AD) and payload (used as ciphertext). Similarly for encryption, the payload is encrypted first, and then the packet number is also encrypted.

Having in-place operations makes this somewhat better because the implementation doesn't need to do the "write packet number plaintext back" operation since it's done automatically. It's not a huge improvement though, so if you'd rather have the operations not be -in-place anyway that can work as well.

Reply:
Yeah, if the API is only out-of-place we'd probably then have to copy it in place after the fact again.